### PR TITLE
[Merged by Bors] - Stageless: prettier cycle reporting

### LINKED ISSUE #7367: Formatting of system schedule cycle detected error is very hard to follow when many systems form a cycle
## Bevy version

#7267, using the new schedule_v3 module.

## Problem

When cycles are detected, the error message is very hard to read:

```rust
schedule contains at least 1 cycle(s) -- cycle(s) found within:
 ---- 0: ["bevy_asset::assets::Assets<bevy_animation::AnimationClip>::asset_event_system", "bevy_asset::assets::Assets<bevy_audio::audio_output::AudioSink>::asset_event_system", "bevy_asset::assets::Assets<bevy_audio::audio_source::AudioSource>::asset_event_system", "bevy_asset::assets::Assets<bevy_gltf::GltfMesh>::asset_event_system", "bevy_asset::assets::Assets<bevy_gltf::GltfPrimitive>::asset_event_system", "bevy_asset::assets::Assets<bevy_gltf::GltfNode>::asset_event_system", "bevy_asset::assets::Assets<bevy_gltf::Gltf>::asset_event_system", "bevy_asset::assets::Assets<bevy_pbr::pbr_material::StandardMaterial>::asset_event_system", "bevy_asset::assets::Assets<bevy_text::font_atlas_set::FontAtlasSet>::asset_event_system", "bevy_asset::assets::Assets<bevy_text::font::Font>::asset_event_system", "bevy_asset::assets::Assets<bevy_sprite::mesh2d::color_material::ColorMaterial>::asset_event_system", "bevy_asset::assets::Assets<bevy_sprite::texture_atlas::TextureAtlas>::asset_event_system", "bevy_asset::assets::Assets<bevy_render::texture::image::Image>::asset_event_system", "bevy_asset::assets::Assets<bevy_render::mesh::mesh::skinning::SkinnedMeshInverseBindposes>::asset_event_system", "bevy_asset::assets::Assets<bevy_render::mesh::mesh::Mesh>::asset_event_system", "bevy_asset::assets::Assets<bevy_render::render_resource::shader::Shader>::asset_event_system", "bevy_asset::assets::Assets<bevy_scene::scene::Scene>::asset_event_system", "bevy_asset::assets::Assets<bevy_scene::dynamic_scene::DynamicScene>::asset_event_system", "bevy_pbr::light::check_light_mesh_visibility", "bevy_pbr::light::update_point_light_frusta", "bevy_pbr::light::update_spot_light_frusta", "bevy_pbr::light::assign_lights_to_clusters", "bevy_pbr::light::update_directional_light_frusta", "bevy_render::view::visibility::check_visibility", "bevy_render::view::visibility::update_frusta<bevy_render::camera::projection::OrthographicProjection>", "bevy_render::view::visibility::update_frusta<bevy_render::camera::projection::PerspectiveProjection>", "bevy_render::view::visibility::update_frusta<bevy_render::camera::projection::Projection>", "bevy_ui::update::update_clipping_system", "bevy_transform::systems::propagate_transforms", "bevy_transform::systems::sync_simple_transforms", "bevy_animation::animation_player", "bevy_audio::audio_output::play_queued_audio_system<bevy_audio::audio_source::AudioSource>", "bevy_pbr::light::add_clusters", "bevy_ui::stack::ui_stack_system", "bevy_ui::flex::flex_node_system", "bevy_ui::widget::image::update_image_calculated_size_system", "bevy_ui::widget::text::text_system", "bevy_text::text2d::update_text2d_layout", "bevy_render::view::visibility::visibility_propagate_system", "bevy_render::camera::camera::camera_system<bevy_render::camera::projection::PerspectiveProjection>", "bevy_render::camera::camera::camera_system<bevy_render::camera::projection::OrthographicProjection>", "bevy_render::camera::camera::camera_system<bevy_render::camera::projection::Projection>", "bevy_winit::system::despawn_window", "bevy_winit::system::changed_window", "bevy_window::system::exit_on_all_closed", "bevy_ecs::scheduling::executor::apply_system_buffers"]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DependencyCycle', crates/bevy_ecs/src/scheduling/schedule.rs:181:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

These don't have the appropriate line breaks in the terminal.

image

## Proposed fixes

1. Use double line breaks for each cycle, and single line breaks for each system.
2. Use short-names by default, with an option to see the full name.
3. List the total number of disconnected cycles at the top of the error message.
4. List the sets that each system is in.

Ambiguity detection recently had similar changes made; check that code for breadcrumbs.
If some of these fixes end up being more involved, submit the easy ones in their own PR.

### PR DESCRIPTION
**Pull request by maniwani**
Graph theory make head hurty. Closes #7367.

Basically, when we topologically sort the dependency graph, we already find its strongly-connected components (a really [neat algorithm][1]). This PR adds an algorithm that can dissect those into simple cycles, giving us more useful error reports.

test: `system_build_errors::dependency_cycle`
```
schedule has 1 before/after cycle(s):
cycle 1: system set 'A' must run before itself
system set 'A'
 ... which must run before system set 'B'
 ... which must run before system set 'A'
```
```
schedule has 1 before/after cycle(s):
cycle 1: system 'foo' must run before itself
system 'foo'
 ... which must run before system 'bar'
 ... which must run before system 'foo'
```

test: `system_build_errors::hierarchy_cycle`
```
schedule has 1 in_set cycle(s):
cycle 1: system set 'A' contains itself
system set 'A'
 ... which contains system set 'B'
 ... which contains system set 'A'
 ```

[1]: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm

### CODE REVIEW

File: crates/bevy_ecs/src/schedule_v3/schedule.rs [hunk 1]
```diff
@@ -961,25 +964,27 @@ impl ScheduleGraph {
         true
     }
 
-    fn report_cycles(&self, strongly_connected_components: &[Vec<NodeId>]) {
-        let components_with_cycles = strongly_connected_components
-            .iter()
-            .filter(|scc| scc.len() > 1)
-            .cloned()
-            .collect::<Vec<_>>();
+    fn report_cycles(&self, cycles: &[Vec<NodeId>], switch: bool) {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule_v3/schedule.rs:967]: The use of `switch` here is pretty unclear to me. Can we make this more explicit?

File: crates/bevy_ecs/src/schedule_v3/schedule.rs [hunk 2]
```diff
@@ -961,25 +964,27 @@ impl ScheduleGraph {
...
+        let tag1 = if switch { "dependencies" } else { "hierarchy" };
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule_v3/schedule.rs:968]: More descriptive variable names please.

File: crates/bevy_ecs/src/schedule_v3/graph_utils.rs [hunk 3]
```diff
@@ -231,3 +231,131 @@ where
...
+    graph: &DiGraphMap<N, ()>,
+    sccs: &[Vec<N>],
+) -> Vec<Vec<N>>
+where
+    N: NodeTrait + Debug,
+{
+    let mut cycles = Vec::new();
+
+    // reporting self-loops as cycles was not part of the original algorithm
+    // for node in graph.nodes() {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule_v3/graph_utils.rs:251]: We probably want this, right? We shouldn't be leaving around commented out code regardless though.

Comment (maniwani) [reply crates/bevy_ecs/src/schedule_v3/graph_utils.rs:251]: Sure, I just commented it out because I was trying to get it to work and loops cannot exist in a schedule, `add_system` and `configure_set` error.

File: crates/bevy_ecs/src/schedule_v3/schedule.rs (hunk 1 above)
Comment (maniwani) [reply crates/bevy_ecs/src/schedule_v3/schedule.rs:967]: Yep.

**Comment by alice-i-cecile** [comment]
Would be nice to get this in for 0.10, but not essential. Removing from the milestone.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/graph_utils.rs [hunk 4]
```diff
@@ -274,3 +274,119 @@ where
...
+pub fn simple_cycles_in_component<N>(graph: &DiGraphMap<N, ()>, scc: &[N]) -> Vec<Vec<N>>
+where
+    N: NodeTrait + Debug,
+{
+    let mut cycles = vec![];
+    let mut sccs = vec![scc.to_vec()];
+
+    while let Some(mut scc) = sccs.pop() {
+        // only look at nodes and edges in this strongly-connected component
+        let mut subgraph = DiGraphMap::new();
```
Comment (DasLixou) [inline crates/bevy_ecs/src/schedule/graph_utils.rs:293]: Can we also set a capacity here?

File: crates/bevy_ecs/src/schedule/graph_utils.rs [hunk 5]
```diff
@@ -274,3 +274,119 @@ where
...
+///
+/// [1]: https://doi.org/10.1137/0204007
+pub fn simple_cycles_in_component<N>(graph: &DiGraphMap<N, ()>, scc: &[N]) -> Vec<Vec<N>>
+where
+    N: NodeTrait + Debug,
+{
+    let mut cycles = vec![];
+    let mut sccs = vec![scc.to_vec()];
+
+    while let Some(mut scc) = sccs.pop() {
```
Comment (DasLixou) [inline crates/bevy_ecs/src/schedule/graph_utils.rs:291]: Why do we have a whole other Vec wrapping around `scc` as a Vec and then popping the `sccs` wrapper in a while-loop when there's always just this one Vec inside?

**Review by DasLixou — APPROVED** [review]
Ok all in all it looks good. I'm not sure if you'll need two different methods just for other error messages but I'll leave it as it is. The DFS looks quite complicated with the dozens of block and unblock stacks, maps, etc. why is this needed?

**Comment by DasLixou** [comment]
And `strongly-connected` means that on directed graphs it’s outgoing, doesn’t it? And does `.neighbours` then also just return those?

File: crates/bevy_ecs/src/schedule/graph_utils.rs (hunk 5 above)
Comment (maniwani) [reply crates/bevy_ecs/src/schedule/graph_utils.rs:291]: Once a node is done being the "root", it's removed from the graph and we have to split the remaining subgraph into new SCCs. Those are pushed to `sccs`.

Comment (DasLixou) [reply crates/bevy_ecs/src/schedule/graph_utils.rs:291]: Oh ok

Comment (DasLixou) [reply crates/bevy_ecs/src/schedule/graph_utils.rs:291]: LOL I missed the last line yeah makes sense

**Comment by maniwani** [comment]
> The algorithm looks quite complicated with the dozens of block and unblock stacks, maps, etc. Why is this needed?

That's just how the algorithm works. I didn't come up with it and I have no clue if a more efficient one exists or if that's even possible. I watched a YT video to get a feel for it. If you're curious about how it works, I'd recommend that.

This might not be the fastest implementation of it, but it only runs in situations where the graph cannot be built (because it contains cycles).

> And strongly-connected means that on directed graphs it’s outgoing, doesn’t it? And does `.neighbours` then also just return those?

A group of nodes in a directed graph are "strongly-connected" if each node in the group has a path to every other node in the group. `neighbours` does returns the outgoing nodes for `DiGraphMap`, but I could be more explicit and use `neighbors_directed`.
